### PR TITLE
#164797157 create article pagination endpoint

### PR DESCRIPTION
**Pull request by igbominadeveloper**
#### What does this PR do?
creates an endpoint for paginating articles retrieved from the database

#### Description of Task to be completed?
- [x] write unit tests
- [x] create endpoint for fetching all articles
- [x] add pagination to the endpoint for limits and page

#### How should this be manually tested?
- Clone the repo,
- Run ```npm install``` to install all dependencies,
- Create .env file using the .env.example as a guide and enter you environment variables
- Run npm test 
- Run npm run start:dev,
- On Postman, send a GET request to localhost:3000/api/v1/articles?limit=10&page=2

#### Any background context you want to provide?
If all articles are fetched at once, this could cause a lot of overhead for the server, hence, why this is needed.
#### What are the relevant pivotal tracker stories? (if applicable)
[#164797157](https://www.pivotaltracker.com/story/show/164797157)

#### Screenshots
Screenshot 2019-04-08 at 9 34 40 AM
Screenshot 2019-04-08 at 7 05 00 AM

### CODE REVIEW

File: src/routes/controllers/articles.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import models from '../../db/models';
```
Comment (Japhethca) [inline src/routes/controllers/articles.js]: This file is redundant.

Comment (igbominadeveloper) [reply src/routes/controllers/articles.js]: I have removed it

File: src/routes/user.route.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { Router } from 'express';
```
Comment (Japhethca) [inline src/routes/user.route.js]: we can remove this file.

File: src/tests/db/controller/user.test.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import chai from 'chai';
```
Comment (Japhethca) [inline src/tests/db/controller/user.test.js]: same as above.

File: src/routes/controllers/articles.controller.js [hunk 4]
```diff
@@ -70,33 +70,67 @@ export async function createArticle(req, res) {
...
+ * @returns {Object} res with articles array if it exists
+ * @returns {Object} res with 404 response if the array is empty
+ */
+export async function getAllArticles(req, res) {
+  const currentPage = req.query.page || 1;
+  const limit = req.query.limit || 10;
+  const offset = limit * (currentPage - 1);
+  try {
+    const articles = await Article.findAndCountAll({
+      attributes: {
```
Comment (Japhethca) [inline src/routes/controllers/articles.controller.js]: we can remove the `attributes` query field.

**Review by Japhethca — CHANGES_REQUESTED** [review]
Just a few comment to address.

File: src/routes/user.route.js (hunk 2 above)
Comment (igbominadeveloper) [reply src/routes/user.route.js]: I have removed it now

File: src/tests/db/controller/user.test.js (hunk 3 above)
Comment (igbominadeveloper) [reply src/tests/db/controller/user.test.js]: I have removed it

File: src/routes/controllers/articles.controller.js (hunk 4 above)
Comment (igbominadeveloper) [reply src/routes/controllers/articles.controller.js]: I have removed it

**Comment by igbominadeveloper** [comment]
@Japhethca changes addressed. Ready for merge

File: package.json [hunk 5]
```diff
@@ -50,7 +48,8 @@
     "db:migrate:testdb": "NODE_ENV=test npm run db:migrate",
     "lint": "node_modules/eslint/bin/eslint.js src --fix",
     "seedall": "NODE_ENV=test sequelize db:seed:all",
-    "coveralls": "cat coverage/lcov.info | coveralls"
+    "coveralls": "cat coverage/lcov.info | coveralls",
+    "db:drop": "sequelize db:drop && sequelize db:create"
```
Comment (Japhethca) [inline package.json]: Am not sure we need this script.

File: src/server.js [hunk 6]
```diff
@@ -63,8 +63,8 @@ app.all('*', (req, res) =>
 );
 
 // finally, let's start our server...
-app.listen(process.env.PORT || 3000, () => {
-  console.log(`Listening on port`);
+const server = app.listen(process.env.PORT || 3000, () => {
```
Comment (Japhethca) [inline src/server.js]: Why this?

Comment (igbominadeveloper) [reply src/server.js]: It was on develop before now, I edited it on my last PR that was merged so, I am just returning it back now

File: package.json (hunk 5 above)
Comment (igbominadeveloper) [reply package.json]: okay. I will get rid of it

File: src/server.js (hunk 6 above)
Comment (igbominadeveloper) [reply src/server.js]: I have removed it now

File: src/routes/controllers/articles.controller.js [hunk 7]
```diff
@@ -162,10 +163,9 @@ export async function getArticle(req, res) {
         article,
       );
     }
-
     return errorResponse(res, 404, 'Article does not exist');
   } catch (error) {
-    return errorResponse(res, 500, 'An error occured', error.message);
+    return errorResponse(res, 500, 'database error', error);
```
Comment (Japhethca) [inline src/routes/controllers/articles.controller.js]: Let's make the error message generic, if the `error` variable contains errors from the database we need to remove it.

File: src/routes/controllers/articles.controller.js [hunk 8]
```diff
@@ -417,3 +418,39 @@ export const recordARead = async (articleId, user = null) => {
...
+    });
+
+    return successResponse(
+      res,
+      200,
+      'Articles successfully retrieved',
+      articles,
+    );
+  } catch (error) {
+    return errorResponse(res, 500, 'An error occured', error.message);
```
Comment (Japhethca) [inline src/routes/controllers/articles.controller.js]: same as above.

**Review by Japhethca — APPROVED** [review]

**Pull request merged**